### PR TITLE
TRITON-1944 Recovery Configurations end-point for CNs using EDAR

### PR DESCRIPTION
**Pull request by kusor**
This one depends into https://github.com/joyent/sdc-cn-agent/pull/25/, though.

### CODE REVIEW

File: lib/endpoints/servers.js [hunk 1]
```diff
@@ -1595,6 +1600,196 @@ Server.nop = function handlerServerNop(req, res, next) {
...
+                    CN_AGENT_VERSION_WITH_RECOVERY_CONFIG)) {
+                cb(new errors.PreconditionFailedError(
+                    'cn-agent (v%s) on server %s does not support '
+                    + '"recovery_config": require at least cn-agent v%s',
+                    ctx.cnAgent.version, req.stash.server.uuid,
+                    CN_AGENT_VERSION_WITH_RECOVERY_CONFIG));
+                return;
+            }
+            self.log.debug(
+                'server %s cn-agent supports "agents_uninstall": v%s',
```
Comment (twhiteman) [inline lib/endpoints/servers.js]: Heh - don't think this is "agents_uninstall".

File: lib/endpoints/servers.js [hunk 2]
```diff
@@ -1754,6 +1949,19 @@ function attachTo(http, app) {
         }),
         Server.factoryReset);
 
+    // Update recovery configuration on the server
+    http.post({
+        path: '/servers/:server_uuid/recovery-config',
+        name: 'ServerRecoveryConfig' },
+        ensure({
+            connectionTimeoutSeconds: 60 * 60,
```
Comment (twhiteman) [inline lib/endpoints/servers.js]: Does 60 minutes seem too long?

File: docs/index.md [hunk 3]
```diff
@@ -1265,15 +1265,16 @@ Initiate the server setup process for a newly started server.
...
+| Param             | Type   | Description                                                  |
+| ----------------- | ------ | ------------------------------------------------------------ |
+| nics              | Object | Nic parameters to update                                     |
+| postsetup_script  | String | Script to run after setup has completed                      |
+| hostname          | String | Hostname to set for the specified server                     |
+| disk_spares       | String | See `man disklayout` spares                                  |
+| disk_width        | String | See `man disklayout` width                                   |
+| disk_cache        | String | See `man disklayout` cache                                   |
+| disk_layout       | String | See `man disklayout` type      (single, mirror, raidz1, ...) |
+| encryption_enabled| String | See `man mkzpool -e`                                         |
```
[the remark this answers is not in the document]
Comment (kusor) [reply docs/index.md]: Done

File: lib/endpoints/servers.js [hunk 4]
```diff
@@ -742,6 +741,8 @@ Server.factoryReset = function handlerServerFactoryReset(req, res, next) {
  * @param {String} disk_cache See `man disklayout` cache
  * @param {String} disk_layout See `man disklayout` type
  *      (single, mirror, raidz1, ...)
+ * @param {String} encryption_enabled Encrypt or not the server zpool. (Boolean
```
[the remark this answers is not in the document]
Comment (kusor) [reply lib/endpoints/servers.js]: English related suggestions always took!

File: lib/endpoints/servers.js [hunk 5]
```diff
@@ -1595,6 +1600,196 @@ Server.nop = function handlerServerNop(req, res, next) {
...
+ *      reported by the Server's sysinfo as staged through the `Zpool Recovery`
+ *      sysinfo property.
+ * @param {String} template pivy-box recovery configuration template to
+ *      be staged into the CN.
+ * @param {String} token the recovery token to stage with the recovery
+ *      configuration.
+ * @param {String} pivtoken GUID of the PIVToken the recovery token is
+ *      associated with.
+ *
+ * @response 200 Ok Install task initiated successfully
```
[the remark this answers is not in the document]
Comment (kusor) [reply lib/endpoints/servers.js]: Done

File: lib/endpoints/servers.js (hunk 1 above)
Comment (kusor) [reply lib/endpoints/servers.js]: /me laughs

File: lib/endpoints/servers.js (hunk 2 above)
Comment (kusor) [reply lib/endpoints/servers.js]: It's actually the default for everything there so I decided to do not modify all of them.

**Pull request merged**